### PR TITLE
Add ability to specify mavenCoord for version range and recommended version

### PR DESCRIPTION
**Pull request by ghost**
This change is [Reviewable](https://reviewable.io/reviews/palantir/sls-packaging/216)

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/sls-packaging](https://github.com/palantir/sls-packaging), @jnewman2501! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy [hunk 1]
```diff
@@ -25,6 +25,15 @@ class ProductDependency {
         isValid()
     }
 
+    def templateMaximumVersionFromMinimumVersion() {
+        if (!minimumVersion) {
+            throw new IllegalArgumentException(
+                    "minimumVersion must be set in order to invoke singleMajorRevMaximumVersion")
```
Comment (ellisjoe) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy]: wrong method name

Comment (ghost) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy]: done

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.groovy [hunk 2]
```diff
@@ -44,11 +53,22 @@ class BaseDistributionExtension {
...
+        productDependency(new ProductDependency(
+                matcher.group("group"),
+                matcher.group("name"),
+                matcher.group("version"),
+                null,
+                recommendedVersion))
     }
 
-    void productDependency(String serviceGroup, String serviceName, String minVersion, String maxVersion, String recommendedVersion) {
+    void productDependency(String serviceGroup, String serviceName, String minVersion, String maxVersion, String recommendedVersion = null) {
```
Comment (markelliot) [inline gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.groovy]: could also do maxVersion = null

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy [hunk 3]
```diff
@@ -35,4 +35,32 @@ class BaseDistributionExtensionTest extends Specification {
         def ex = thrown IllegalArgumentException
         ex.message == "Invalid product type 'foobar' specified; supported types: [service.v1, asset.v1]."
     }
+
+    def "productDependencies from maven coordinate and no recommended version"() {
+        when:
+        def ext = new BaseDistributionExtension(null)
+        ext.productDependency("group:name:1.2.3:classifier@tgz")
```
Comment (markelliot) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy]: would like to see some examples without classifier and without artifact type

**Review by markelliot — CHANGES_REQUESTED** [review]

File: gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.groovy (hunk 2 above)
Comment (ghost) [reply gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.groovy]: done

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy (hunk 3 above)
Comment (ghost) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy]: done

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy [hunk 4]
```diff
@@ -35,4 +35,55 @@ class BaseDistributionExtensionTest extends Specification {
         def ex = thrown IllegalArgumentException
         ex.message == "Invalid product type 'foobar' specified; supported types: [service.v1, asset.v1]."
     }
+
+    def "productDependencies from invalid maven coordinat"() {
```
Comment (markelliot) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy]: coordinat -> coordinate

**Review by markelliot — APPROVED** [review]
good besides nit

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy (hunk 4 above)
Comment (ghost) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy]: done!

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy [hunk 5]
```diff
@@ -35,4 +35,55 @@ class BaseDistributionExtensionTest extends Specification {
...
+        ext.serviceDependencies.get(0).productName == "name"
+        ext.serviceDependencies.get(0).minimumVersion == "1.2.3"
+        ext.serviceDependencies.get(0).maximumVersion == "1.x.x"
+        ext.serviceDependencies.get(0).recommendedVersion == null
+    }
+
+    def "productDependencies from maven coordinate with all fields and no recommended version"() {
+        when:
+        def ext = new BaseDistributionExtension(null)
+        ext.productDependency("group:name:1.2.3:classifier@tgz")
```
Comment (uschi2000) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy:65]: do we need to allow this?

Comment (uschi2000) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy:65]: do we need to support this syntax?

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy [hunk 6]
```diff
@@ -35,4 +35,55 @@ class BaseDistributionExtensionTest extends Specification {
         def ex = thrown IllegalArgumentException
         ex.message == "Invalid product type 'foobar' specified; supported types: [service.v1, asset.v1]."
     }
+
+    def "productDependencies from invalid maven coordinate"() {
+        when:
+        def ext = new BaseDistributionExtension(null)
+        ext.productDependency("group:name")
```
Comment (uschi2000) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy:42]: too bad this doesn't plug into the gradle's version resolution mechanism , i.e., it doesn't work with nebula and versions.props.

Comment (uschi2000) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy:42]: too bad this doesn't plug into the gradle's version resolution mechanism , i.e., it doesn't work with nebula and versions.props.

**Review by uschi2000 — COMMENTED** [review]
README needs an update.

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy (hunk 6 above)
Comment (markelliot) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy:42]: with this, you would call, for instance:

    distribution {
    ...
        productDependency recommend('group:name')
    }

Comment (markelliot) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy:42]: and this decomp means this plugin doesn't require nebula

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy (hunk 5 above)
Comment (markelliot) [reply gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/BaseDistributionExtensionTest.groovy:65]: don't think it's harmful

File: gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/ProductDependencyTest.groovy [hunk 7]
```diff
@@ -26,4 +26,21 @@ class ProductDependencyTest extends Specification {
...
+
+    def 'default maximumVersion'() {
+        when:
+        def dep = new ProductDependency("", "", "1.2.3", null, "1.2.4")
+
+        then:
+        dep.maximumVersion == "1.x.x"
+    }
+
+    def 'non-deafult maximumVersion'() {
```
Comment (ashrayjain) [inline gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/ProductDependencyTest.groovy:38]: typo?

**Pull request merged**